runtime/js/telemetry.js

```diff
@@ -0,0 +1,378 @@
+        const span = spans[i];
+        const context = span.spanContext();
+        submit({
+          spanId: context.spanId,
+          traceId: context.traceId,
+          traceFlags: context.traceFlags,
+          name: span.name,
+          kind: span.kind,
+          parentSpanId: span.parentSpanId,
+          startTime: hrToSecs(span.startTime),
+          endTime: hrToSecs(span.endTime),
+          status: span.status,
+          attributes: span.attributes,
+        });
+      }
+      resultCallback({ code: 0 });
+    } catch (error) {
+      resultCallback({ code: 1, error });
+    }
+  }
+
+  async shutdown() {}
+
+  async forceFlush() {}
+}
+
+// SPAN_KEY matches symbol in otel-js library
+const SPAN_KEY = SymbolFor("OpenTelemetry Context Key SPAN");
+
+// Context tracker compatible with otel-js api
+class Context {
+  #data = new SafeMap();
+
+  constructor(data) {
+    this.#data = data ? new SafeMap(data) : new SafeMap();
+  }
+
+  getValue(key) {
+    return this.#data.get(key);
+  }
+
+  setValue(key, value) {
+    const c = new Context(this.#data);
+    c.#data.set(key, value);
+    return c;
+  }
+
+  deleteValue(key) {
+    const c = new Context(this.#data);
+    c.#data.delete(key);
+    return c;
+  }
+}
+
+const ROOT_CONTEXT = new Context();
+
+// Context manager for opentelemetry js library
+class ContextManager {
+  active() {
+    return CURRENT.get() ?? ROOT_CONTEXT;
+  }
+
+  with(context, fn, thisArg, ...args) {
+    const ctx = CURRENT.enter(context);
+    try {
+      return ReflectApply(fn, thisArg, args);
+    } finally {
+      setAsyncContext(ctx);
+    }
+  }
+
+  bind(context, f) {
+    return (...args) => {
+      const ctx = CURRENT.enter(context);
+      try {
+        return ReflectApply(f, thisArg, args);
+      } finally {
+        setAsyncContext(ctx);
+      }
+    };
+  }
+
+  enable() {
+    return this;
+  }
+
+  disable() {
+    return this;
+  }
+}
+
+function otelLog(message, level) {
+  let traceId = "";
+  let spanId = "";
+  let traceFlags = 0;
+  const span = Span.current();
+  if (span) {
+    if (span.spanContext) {
+      const context = span.spanContext();
+      traceId = context.traceId;
+      spanId = context.spanId;
+      traceFlags = context.traceFlags;
+    } else {
+      traceId = span.traceId;
+      spanId = span.spanId;
+      traceFlags = span.traceFlags;
+    }
+  }
+  return op_otel_log(message, level, traceId, spanId, traceFlags);
+}
+
+const otelConsoleConfig = {
+  ignore: 0,
+  capture: 1,
+  replace: 2,
+};
+
+export function bootstrap(config) {
+  if (config.length === 0) return;
+  const { 0: consoleConfig } = config;
+
+  TRACING_ENABLED = true;
+
+  switch (consoleConfig) {
+    case otelConsoleConfig.capture:
+      core.wrapConsole(globalThis.console, new Console(otelLog));
+      break;
+    case otelConsoleConfig.replace:
+      ObjectDefineProperty(
+        globalThis,
+        "console",
+        core.propNonEnumerable(new Console(otelLog)),
+      );
```

Comment:
It should probably still be wired into the original V8 console so that inspector console stays working.

Reply:
that happens in `99_main.js`.